src/lexer.rs

```diff
@@ -151,6 +151,57 @@ impl<'a> Lexer<'a> {
         Token::Ident(s, ss.as_str().into())
     }
 
+    /// Simulate reading from standard input after a statement like `COPY ... FROM STDIN;`.
+    /// First skips space characters and optionally one NL.
+    /// Then consumes until NL '\' '.' NL is encountered, or until EOF.
+    /// The trailing '\' '.' NL is consumed but not returned.
+    pub fn read_from_stdin(&mut self) -> (&'a str, Span) {
+        enum State {
+            Init,
+            Nl,
+            NlBackslash,
+            NlBackslashDot,
+        }
+        // Skip optional spaces followed by one newline.
+        let (start, mut state) = 's: loop {
+            while let Some(&(i, c)) = self.chars.peek() {
+                if c == '\n' {
+                    self.chars.next().unwrap();
+                    break 's (i + 1, State::Nl);
+                }
+                if c.is_ascii_whitespace() {
+                    self.chars.next().unwrap();
+                    continue;
+                }
+                break 's (i, State::Init);
+            }
+            let span = self.src.len()..self.src.len();
+            return (self.s(span.clone()), span);
+        };
+        while let Some((i, c)) = self.chars.next() {
+            state = if matches!(state, State::Init) && c == '\n' {
+                State::Nl
+            } else if matches!(state, State::Nl) && c == '\\' {
+                State::NlBackslash
+            } else if matches!(state, State::NlBackslash) && c == '.' {
+                State::NlBackslashDot
+            } else if matches!(state, State::NlBackslashDot) && c == '\n' {
+                let span = start..(i - 2);
+                return (self.s(span.clone()), span);
+            } else {
+                State::Init
```

Comment:
There would be less branching if we use the program counter as the state, perhaps something like the following untested code:
```rust
pub fn read_from_stdin(&mut self) -> (&'a str, Span) {
    // Skip optional whitespace
    while self.chars.peek().iter().any(|(_, c)| c.is_ascii_whitespace()) {
        self.chars.next().unwrap();
    }
    let start = match self.chars.peek() {
        Some((i, '\n')) => i+1,
        Some((i, _)) => i,
        None => {
            let span = self.src.len()..self.src.len();
            return (self.s(span.clone()), span);
        }
    };
    while let Some((i, c)) = self.chars.next() {
        if c != '\n' {
            continue
        }
        if !matches!(self.chars.peek(), Some(_, '\\') {
            continue;
        }
        self.chars.next().unwrap();
        if !matches!(self.chars.peek(), Some(_, '.') {
            continue;
        }
        self.chars.next().unwrap();
        match self.chars.peek() {
            Some((i, '\n')) => {
                self.chars.next();
                let span = start..(i - 2);
                return (self.s(span.clone()), span);
            }
            None => {
                // Allow data to end with NL '\' '.' without an extra NL,
                // and skip the last '\' '.'.
                let span = start..(self.src.len() - 2);
                return (self.s(span.clone()), span);
            }
            _ => continue
        }
        self.chars.next().unwrap();
    }
    let span = start..self.src.len();
    return (self.s(span.clone()), span);    
}
```